### PR TITLE
External executor

### PR DESCRIPTION
**Pull request by nc6**
Module to run external tasks against a content store.

### CODE REVIEW

**Comment by nc6** [comment]
Result of the testing here:
```
sudo nixos-container create redis --config-path redis.nix
sudo nixos-container start redis
stack exec ffexecutord -- --storePath /tmp/tmp.i54QL7iKhU --redisHost 10.233.2.2 --redisPort 6379

stack test:
ContentHash {unContentHash = 2e2d72f9b1b518d635e3348bc7fab3b0b07c20c50197375b00a704e6565a25a9}

[nc@varda:~/topos/pfizer/funflow]$ l /tmp/tmp.i54QL7iKhU/
total 116K
dr-xr-xr-x   3 nc   users 4.0K Nov  7 15:43 .
drwxrwxrwt 135 root root  104K Nov  7 15:43 ..
dr-xr-xr-x   2 nc   users 4.0K Nov  7 15:43 Li1y-bG1GNY14zSLx_qzsLB8IMUBlzdbAKcE5lZaJak

[nc@varda:~/topos/pfizer/funflow]$ cat /tmp/tmp.i54QL7iKhU/Li1y-bG1GNY14zSLx_qzsLB8IMUBlzdbAKcE5lZaJak/out 
Hello World

[nc@varda:~/topos/pfizer/funflow]$ stack test
funflow-0.1.0: test (suite: test-funflow)

ContentHash {unContentHash = 2e2d72f9b1b518d635e3348bc7fab3b0b07c20c50197375b00a704e6565a25a9}

[nc@varda:~/topos/pfizer/funflow]$ l /tmp/tmp.i54QL7iKhU/
total 116K
dr-xr-xr-x   3 nc   users 4.0K Nov  7 15:43 .
drwxrwxrwt 135 root root  104K Nov  7 15:43 ..
dr-xr-xr-x   2 nc   users 4.0K Nov  7 15:43 Li1y-bG1GNY14zSLx_qzsLB8IMUBlzdbAKcE5lZaJak
```

File: src/Control/FunFlow/External/Executor.hs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import           System.Exit                          (ExitCode (..))
+import           System.FilePath                      ((</>))
+import           System.IO                            (IOMode (..), openFile)
+import           System.Process
+
+data ExecutionResult =
+    -- | The result already exists in the store and there is no need
+    --   to execute. This is also returned if the job is already running
+    --   elsewhere.
+    Cached
```
Comment (aherrmann) [inline src/Control/FunFlow/External/Executor.hs]: > This is also returned if the job is already running elsewhere.

That sounds dangerous.

File: src/Control/FunFlow/External/Executor.hs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      afterFailure t i = Failed (ExecutionInfo executor t) i
+
+  forever $ do
+    mtask <- popTask hook executor
+    case mtask of
+      Nothing -> return ()
+      Just task -> do
+        res <- execute store task
+        updateTaskStatus hook (task ^. tdOutput) $ case res of
+          Cached      -> fromCache
```
Comment (aherrmann) [inline src/Control/FunFlow/External/Executor.hs]: Won't this mark tasks that are still running as already completed?

Per

> This is also returned if the job is already running elsewhere.

File: src/Control/FunFlow/External/Executor.hs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        mp <- try $ createProcess $ procSpec out
+        case mp of
+          Left (_ex :: IOException) -> do
+            removeFailed store (td ^. tdOutput)
+            return $ Failure (diffTimeSpec start start) 2
+          Right (_, _, _, ph) -> do
+            exitCode <- waitForProcess ph
+            end <- getTime Monotonic
+            case exitCode of
+              ExitSuccess   -> do
```
Comment (aherrmann) [inline src/Control/FunFlow/External/Executor.hs]: This is the only path that leads to a store entry being marked as complete, right?
But how to handle the case where ffexecutord is restarted while external processes are running?
It looks like currently `execute` would mark these external tasks as `Cached` and execute loop would then mark the status as `Completed`.

File: src/Control/FunFlow/External/Executor.hs (hunk 2 above)
Comment (nc6) [reply src/Control/FunFlow/External/Executor.hs]: Yes. Hmm, what's a good way to handle already running jobs? Ideally we shouldn't get them....

File: src/Control/FunFlow/External/Executor.hs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+-- | Execute tasks forever
+executeLoop :: forall c. Coordinator c
+            => c
+            -> Config c
+            -> FilePath
+            -> IO ()
+executeLoop _ cfg sroot = do
+  hook :: Hook c <- initialise cfg
+  executor <- Executor <$> getHostName
+  store <- initialize sroot
```
[the remark this answers is not in the document]
Comment (nc6) [reply src/Control/FunFlow/External/Executor.hs]: I like qualified imports, I'll move to that.

**Pull request merged**